File: opensource/kernel/src/main/java/com/sos/scheduler/engine/main/SchedulerThreadController.java
package com.sos.scheduler.engine.main;

import com.google.common.collect.ImmutableList;
import com.sos.scheduler.engine.common.sync.ThrowableMailbox;
import com.sos.scheduler.engine.common.time.Stopwatch;
import com.sos.scheduler.engine.eventbus.SchedulerEventBus;
import com.sos.scheduler.engine.kernel.Scheduler;
import com.sos.scheduler.engine.kernel.settings.CppSettings;
import org.joda.time.Duration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.concurrent.TimeUnit;

import static com.google.common.base.Preconditions.checkState;

/** Steuert den {@link SchedulerThread}. */
public class SchedulerThreadController implements SchedulerController {
    private static final Logger logger = LoggerFactory.getLogger(SchedulerThreadController.class);
    private static final Duration terminationTimeout = Duration.standardSeconds(5);

    private final String name;
    private final SchedulerEventBus eventBus = new SchedulerEventBus();
    private boolean isStarted = false;
    private final ThrowableMailbox<Throwable> throwableMailbox = new ThrowableMailbox<Throwable>();
    private final SchedulerThreadControllerBridge controllerBridge;
    private final SchedulerThread thread;

    public SchedulerThreadController(String name, CppSettings cppSettings) {
        this.name = name;
        controllerBridge = new SchedulerThreadControllerBridge(this, eventBus, cppSettings);
        thread = new SchedulerThread(controllerBridge);
    }

    public final void loadModule(File cppModuleFile) {
        SchedulerThread.loadModule(cppModuleFile);
    }

    @Override public final void startScheduler(Iterable<String> args) {
        checkIsNotStarted();
        controllerBridge.start();
        thread.startThread(ImmutableList.copyOf(args));
        isStarted = true;
    }

    @Override public final void close() {
        Stopwatch stopwatch = new Stopwatch();
        terminateScheduler();
        if (!tryJoinThread(terminationTimeout)) {
            logger.warn("Still waiting for JobScheduler termination ("+terminationTimeout+") ...");
            tryJoinThread(new Duration(Long.MAX_VALUE));
            logger.info("JobScheduler has been terminated after "+stopwatch);
        }
        controllerBridge.close();
        eventBus.dispatchEvents();  // Thread-sicher, weil der Scheduler-Thread beendet ist, also selbst kein dispatchEvents() mehr aufrufen kann.
        throwableMailbox.throwUncheckedIfSet();
    }

    public final Scheduler waitUntilSchedulerState(SchedulerState s) {
        try {
            checkIsStarted();
            Scheduler scheduler = controllerBridge.waitUntilSchedulerState(s);
            throwableMailbox.throwUncheckedIfSet();
            return scheduler;
        }
        catch (InterruptedException x) { throw new RuntimeException(x); }
    }

    @Override public final void terminateAfterException(Throwable t) {
        throwableMailbox.setIfFirst(t);
        terminateScheduler();
    }

    final void setThrowable(Throwable t) {
        throwableMailbox.setIfFirst(t);
    }

    @Override public final void terminateScheduler() {
        controllerBridge.terminate();
    }

    @Override public final boolean tryWaitForTermination(Duration timeout) {
        checkIsStarted();
        boolean result = tryJoinThread(timeout);
        throwableMailbox.throwUncheckedIfSet();
        return result;
    }

    private boolean tryJoinThread(Duration timeout) {
        try {
            if (timeout.getMillis() == Long.MAX_VALUE)  thread.join();
            else TimeUnit.MILLISECONDS.timedJoin(thread, timeout.getMillis());
        }
        catch (InterruptedException x) { throw new RuntimeException(x); }
        return !thread.isAlive();
    }

    public void checkIsNotStarted() {
        checkState(!isStarted, "Scheduler has already been started");
    }

    private void checkIsStarted() {
        checkState(isStarted, "Scheduler has not been started");
    }

    public final boolean isStarted() {
        return isStarted;
    }

    @Override public final int exitCode() {
        return thread.exitCode();
    }

    @Override public final SchedulerEventBus getEventBus() {
        return eventBus;
    }

    public String getName() {
        return name;
    }
}

File: engine-kernel/src/main/scala/com/sos/scheduler/engine/main/SchedulerThreadController.scala
package com.sos.scheduler.engine.main

import com.sos.jobscheduler.common.scalautil.Logger
import com.sos.jobscheduler.common.time.ScalaTime._
import com.sos.jobscheduler.common.time.Stopwatch
import com.sos.scheduler.engine.common.sync.ThrowableMailbox
import com.sos.scheduler.engine.eventbus.SchedulerEventBus
import com.sos.scheduler.engine.kernel.Scheduler
import com.sos.scheduler.engine.kernel.settings.CppSettings
import com.sos.scheduler.engine.main.SchedulerThreadController._
import java.io.File
import java.time.Duration
import java.util.concurrent.TimeUnit
import org.scalactic.Requirements._
import scala.collection.JavaConversions._

/**
 * Steuert den [[SchedulerThread]].
 *
 * @author Joacim Zschimmer
 */
final class SchedulerThreadController(val name: String, cppSettings: CppSettings) extends SchedulerController {
  private val _eventBus = new SchedulerEventBus
  private val throwableMailbox: ThrowableMailbox[Throwable] = new ThrowableMailbox[Throwable]
  private val controllerBridge = new SchedulerThreadControllerBridge(this, eventBus, cppSettings)
  private val thread = new SchedulerThread(controllerBridge)
  private var _isStarted: Boolean = false

  def injector = controllerBridge.injector

  def loadModule(cppModuleFile: File): Unit =
    SchedulerThread.loadModule(cppModuleFile)

  def startScheduler(args: String*): Unit = {
    checkIsNotStarted()
    controllerBridge.start()
    thread.startThread(args)
    _isStarted = true
  }

  def close(): Unit = {
    val stopwatch = new Stopwatch
    terminateScheduler()
    if (!tryJoinThread(terminationTimeout)) {
      logger.warn(s"Still waiting for JobScheduler termination (${terminationTimeout.pretty}) ...")
      tryJoinThread(Duration.ofMillis(Long.MaxValue))
      logger.info(s"JobScheduler has been terminated after $stopwatch")
    }
    controllerBridge.close()
    eventBus.dispatchEvents()  // Thread-sicher, weil der Scheduler-Thread beendet ist, also selbst kein dispatchEvents() mehr aufrufen kann.
    throwableMailbox.throwUncheckedIfSet()
  }

  def waitUntilSchedulerState(s: BridgeState): Scheduler = {
    checkIsStarted()
    val scheduler = controllerBridge.waitUntilSchedulerState(s)
    throwableMailbox.throwUncheckedIfSet()
    scheduler
  }

  def terminateAfterException(t: Throwable): Unit = {
    throwableMailbox.setIfFirst(t)
    terminateScheduler()
  }

  private[main] def setThrowable(t: Throwable): Unit =
    throwableMailbox.setIfFirst(t)

  def terminateScheduler(): Unit = controllerBridge.terminate()

  def tryWaitForTermination(timeout: Duration): Boolean = {
    checkIsStarted()
    val result = tryJoinThread(timeout)
    throwableMailbox.throwUncheckedIfSet()
    result
  }

  private def tryJoinThread(timeout: Duration): Boolean = {
    if (timeout.toMillis == Long.MaxValue) {
      thread.join()
    } else {
      TimeUnit.MILLISECONDS.timedJoin(thread, timeout.toMillis)
    }
    !thread.isAlive
  }

  def checkIsNotStarted(): Unit = requireState(!isStarted, "Scheduler has already been started")

  private def checkIsStarted(): Unit = requireState(isStarted, "Scheduler has not been started")

  def isStarted = _isStarted

  def exitCode: Int = thread.exitCode

  def eventBus = _eventBus

  def getName = name
}

object SchedulerThreadController {
  private val logger = Logger(getClass)
  private val terminationTimeout = 5.s
}
